<html>
  <head>
    <title>Help</title>
  </head>
  <body>
    <h1>Pyrosequence Annotation and Categorization Tool</h1>
    <br><a href="#Parsing">Parsing</a>
    <br><a href="#Database">Viewing Database Results</a>
    <br><a href="#Taxonomies">View Taxonomy Results</a>
    <br><a href="#PieCharts">Pie Charts</a>
    <br>
    <h2>Summary</h2>
    <p>The Pyrosequence Annotation and Categorization Tool, or PACT, was created to automate
    the interpretation of results from large sequence similarity searches, particulary from BLAST.</p>
    <a name="Parsing">
    <h3>Parsing</h3>
    <br>
    <p>To start an analysis with PACT, you need a BLAST output file and the corresponding FASTA sequence file that produced the alignments. PACT will open with the Parser menu, but you can also click "File" and then "New Parser." In the section "Input  Files" enter a name for the parsing job and find the BLAST and FASTA files. These fields are mandatory.
<ul>
<li> Classifications. In addition to the input files, PACT allows you to specify how you would like the hit information of the BLAST file to be categorized according to a classification scheme of your choosing. PACT has a very simple file format for defining classifications. To create a classification file, open a text file and type a title on the first line. To define an identifier, type the pound sign, and then the identifier. Type each item of the identifier below. For example, an identifier in a "Virus Hosts" classificiation file may be "#Humans", and a virus below it may be "ebola." For a new identifier, repeat.
</li>
<li> Flag Files. The input flag file is simply a text file with a list of items to search for in the parsing. While parsing, PACT will print hit and FASTA files for hits that match any of the flag items. These hit and FASTA files will go into a folder with the same title as the flag input file. The title of the flag file must be specified on its very first line.
</li>
<li> Taxonomy. Currently, PACT only bases taxonomic information on the NCBI Taxonomy, which is incomplete and unofficial (but still very good). There are two options for the source of this information. PACT can obtain a hit's taxonomic rank directly from NCBI's <a href="ftp://ftp.ncbi.nih.gov/pub/taxonomy/">website</a>, which requires an internet connection, or it can search files on your computer downloaded from that website. The advantages of using the local files are that you can still categorize by taxonomy without a connection and that the search is a bit faster.  The disadvantage is that all of the hits may not necessarily be found to belong to a taxonomy.  This is because the search on the local files is based on keywords, not on identification number. To choose the source of your taxonomic search, choose "Connection" or "Local Files" in the "Source" drop-down bar (if left blank, PACT will not search by taxonomy at all).
</li>
<li> Parameters. The two parameters for parsing are the bit score and e-value.  If a hit's bit score is lower than the specified value, or if its e-value is higher, PACT will skip it while parsing.
</li>
<li> Output. You must specify an output folder where the results of the parsing will be placed.  Every parsing, regardless of whether the text file option is checked,
will produce a folder called "No Hits" and a summary text file.  The No Hits folder will contain a FASTA file with every hit that did not produce an alignment in the
BLAST output file.  The summary text file contains all highest-scoring hits for each hit sequence, followed by the total number discovered. If the user has defined a taxonomy, classification, or flag files option, the results of those searches will also appear in the output folder.
<br>
If the text files option is checked, PACT will create a folder for each hit containing a FASTA file with the sequences that produced the hit alignment and a text file with information on the hit. If this choice is selected in conjunction with the NCBI taxonomy option, the folders will be nested in a tree structure based on the taxonomy.
</li> 
</ul>
<br>Queuing. You can place several parsing jobs on a queue to be executed all at one time. To add a job to the queue, just press the "add" button while making sure the mandatory fields are filled. To edit a parsing job on the queue, you must make sure that job is highlighted in the queue box. </p>
    <a name="Database">
    <h3>Viewing Database Results</h3>
    <a name="Taxonomies">
    <h3>Viewing Taxonomy Results</h3>
    <p>The summary text files produced from the taxonomy results of a parsing can also be reproduced, if the taxonomy xml file was kept. To reproduce the summary file, go to "View->Tree." Click on "Browse," and then find your taxonomy xml file. A list of nodes in the taxonomy tree structure will appear. If you choose one of these nodes, only the data from this node's child nodes will appear. You can use choose to display only a certain taxonomy level, e.g. "species", or "phylum." Click on the drop-down list under "Select Level" to do this. If multiple taxonomies are added to the viewing list (located on the far right),
    the trees and each node's value will be combined into one tree. To edit </p>
    <a name="PieCharts">
    <h3>Viewing Pie Charts</h3>
    <p>PACT allows you to view pie charts generated from the xml files produced from the results of parsings with classification and/or taxonomy input.</p>
  </body>
</html>
